File: sdks/python/kaivue/services/retention.py
"""Retention policy service client."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from kaivue.models.retention import RetentionPolicy
from kaivue.services.base import BaseService


class RetentionService(BaseService):
    """CRUD operations for retention policies."""

    def create(
        self,
        name: str,
        retention_days: int,
        *,
        description: str = "",
        max_bytes: int = 0,
        event_retention_days: int = 0,
    ) -> RetentionPolicy:
        body: Dict[str, Any] = {
            "name": name,
            "retention_days": retention_days,
        }
        if description:
            body["description"] = description
        if max_bytes:
            body["max_bytes"] = max_bytes
        if event_retention_days:
            body["event_retention_days"] = event_retention_days
        resp = self._post("/v1/retention-policies", body)
        return self._parse(RetentionPolicy, resp["policy"])

    def get(self, policy_id: str) -> RetentionPolicy:
        resp = self._get(f"/v1/retention-policies/{policy_id}")
        return self._parse(RetentionPolicy, resp["policy"])

    def update(
        self,
        policy_id: str,
        *,
        name: Optional[str] = None,
        description: Optional[str] = None,
        retention_days: Optional[int] = None,
        max_bytes: Optional[int] = None,
        event_retention_days: Optional[int] = None,
    ) -> RetentionPolicy:
        body: Dict[str, Any] = {"id": policy_id}
        mask: List[str] = []
        if name is not None:
            body["name"] = name
            mask.append("name")
        if description is not None:
            body["description"] = description
            mask.append("description")
        if retention_days is not None:
            body["retention_days"] = retention_days
            mask.append("retention_days")
        if max_bytes is not None:
            body["max_bytes"] = max_bytes
            mask.append("max_bytes")
        if event_retention_days is not None:
            body["event_retention_days"] = event_retention_days
            mask.append("event_retention_days")
        body["update_mask"] = mask

        resp = self._patch(f"/v1/retention-policies/{policy_id}", body)
        return self._parse(RetentionPolicy, resp["policy"])

    def delete(self, policy_id: str) -> None:
        self._delete(f"/v1/retention-policies/{policy_id}")

    def list(
        self, *, page_size: int = 50, cursor: Optional[str] = None
    ) -> List[RetentionPolicy]:
        params: Dict[str, Any] = {"page_size": page_size}
        if cursor:
            params["cursor"] = cursor
        resp = self._get("/v1/retention-policies", **params)
        return self._parse_list(RetentionPolicy, resp.get("policies", []))

    def apply(self, policy_id: str, camera_ids: List[str]) -> RetentionPolicy:
        """Apply a retention policy to one or more cameras."""
        body = {"policy_id": policy_id, "camera_ids": camera_ids}
        resp = self._post(f"/v1/retention-policies/{policy_id}/apply", body)
        return self._parse(RetentionPolicy, resp["policy"])

File: sdks/go/kaivue/retention.go
package kaivue

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
	"strconv"
)

// RetentionService handles retention policy CRUD operations.
type RetentionService struct {
	client *Client
}

// CreateRetentionPolicyRequest is the input for creating a retention policy.
type CreateRetentionPolicyRequest struct {
	Name               string `json:"name"`
	Description        string `json:"description,omitempty"`
	RetentionDays      int    `json:"retention_days"`
	MaxBytes           int64  `json:"max_bytes,omitempty"`
	EventRetentionDays int    `json:"event_retention_days,omitempty"`
}

// UpdateRetentionPolicyRequest is the input for updating a retention policy.
type UpdateRetentionPolicyRequest struct {
	ID                 string   `json:"id"`
	Name               string   `json:"name,omitempty"`
	Description        string   `json:"description,omitempty"`
	RetentionDays      *int     `json:"retention_days,omitempty"`
	MaxBytes           *int64   `json:"max_bytes,omitempty"`
	EventRetentionDays *int     `json:"event_retention_days,omitempty"`
	UpdateMask         []string `json:"update_mask"`
}

// ListRetentionPoliciesRequest is the input for listing retention policies.
type ListRetentionPoliciesRequest struct {
	PageSize int    `json:"page_size,omitempty"`
	Cursor   string `json:"cursor,omitempty"`
}

// ListRetentionPoliciesResponse wraps a page of retention policies.
type ListRetentionPoliciesResponse struct {
	Policies   []RetentionPolicy `json:"policies"`
	NextCursor string            `json:"next_cursor"`
}

// ApplyRetentionPolicyRequest applies a policy to cameras.
type ApplyRetentionPolicyRequest struct {
	PolicyID  string   `json:"policy_id"`
	CameraIDs []string `json:"camera_ids"`
}

// Create creates a new retention policy.
func (s *RetentionService) Create(ctx context.Context, req *CreateRetentionPolicyRequest) (*RetentionPolicy, error) {
	body, err := s.client.post(ctx, "/v1/retention-policies", req)
	if err != nil {
		return nil, err
	}
	var resp struct {
		Policy RetentionPolicy `json:"policy"`
	}
	if err := json.Unmarshal(body, &resp); err != nil {
		return nil, fmt.Errorf("kaivue: unmarshal retention policy: %w", err)
	}
	return &resp.Policy, nil
}

// Get retrieves a retention policy by ID.
func (s *RetentionService) Get(ctx context.Context, id string) (*RetentionPolicy, error) {
	body, err := s.client.get(ctx, "/v1/retention-policies/"+id, nil)
	if err != nil {
		return nil, err
	}
	var resp struct {
		Policy RetentionPolicy `json:"policy"`
	}
	if err := json.Unmarshal(body, &resp); err != nil {
		return nil, fmt.Errorf("kaivue: unmarshal retention policy: %w", err)
	}
	return &resp.Policy, nil
}

// Update updates a retention policy.
func (s *RetentionService) Update(ctx context.Context, req *UpdateRetentionPolicyRequest) (*RetentionPolicy, error) {
	body, err := s.client.patch(ctx, "/v1/retention-policies/"+req.ID, req)
	if err != nil {
		return nil, err
	}
	var resp struct {
		Policy RetentionPolicy `json:"policy"`
	}
	if err := json.Unmarshal(body, &resp); err != nil {
		return nil, fmt.Errorf("kaivue: unmarshal retention policy: %w", err)
	}
	return &resp.Policy, nil
}

// Delete deletes a retention policy.
func (s *RetentionService) Delete(ctx context.Context, id string) error {
	_, err := s.client.delete(ctx, "/v1/retention-policies/"+id, nil)
	return err
}

// List lists retention policies.
func (s *RetentionService) List(ctx context.Context, req *ListRetentionPoliciesRequest) (*ListRetentionPoliciesResponse, error) {
	params := url.Values{}
	if req.PageSize > 0 {
		params.Set("page_size", strconv.Itoa(req.PageSize))
	} else {
		params.Set("page_size", "50")
	}
	if req.Cursor != "" {
		params.Set("cursor", req.Cursor)
	}

	body, err := s.client.get(ctx, "/v1/retention-policies", params)
	if err != nil {
		return nil, err
	}
	var resp ListRetentionPoliciesResponse
	if err := json.Unmarshal(body, &resp); err != nil {
		return nil, fmt.Errorf("kaivue: unmarshal retention policies list: %w", err)
	}
	return &resp, nil
}

// Apply applies a retention policy to one or more cameras.
func (s *RetentionService) Apply(ctx context.Context, req *ApplyRetentionPolicyRequest) (*RetentionPolicy, error) {
	body, err := s.client.post(ctx, "/v1/retention-policies/"+req.PolicyID+"/apply", req)
	if err != nil {
		return nil, err
	}
	var resp struct {
		Policy RetentionPolicy `json:"policy"`
	}
	if err := json.Unmarshal(body, &resp); err != nil {
		return nil, fmt.Errorf("kaivue: unmarshal retention policy: %w", err)
	}
	return &resp.Policy, nil
}
